File: 133-clone-graph/133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
unordered_map<Node *, Node *> oldToNew;
public:
    Node* cloneGraph(Node* node) {
        if (!node) return NULL;
        if (oldToNew.find(node) != oldToNew.end())
            return oldToNew[node];
        Node *nn = new Node(node->val);
        oldToNew[node] = nn;
        for (auto &nei : node->neighbors) {
            Node *neiNodes = cloneGraph(nei);
            nn->neighbors.push_back(neiNodes);
        }
        return nn;
    }
};

File: 133-clone-graph/133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if node == None:
            return
        oldToNew = {}
        def helper(node):
            if node in oldToNew:
                return oldToNew[node]
            copy = Node(node.val)
            oldToNew[node] = copy
            for nei in node.neighbors:
                copy.neighbors.append(helper(nei))
            return copy
        helper(node)
        return oldToNew[node]